core/ipc/sink/aws-sqs-impl/src/main/java/org/opennms/core/ipc/sink/aws/sqs/AwsRemoteMessageDispatcherFactory.java

```diff
@@ -0,0 +1,199 @@
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.core.ipc.sink.aws.sqs;
+
+import java.io.IOException;
+import java.util.Dictionary;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+
+import javax.jms.JMSException;
+import javax.jms.MessageProducer;
+import javax.jms.Session;
+import javax.jms.TextMessage;
+
+import org.opennms.core.ipc.sink.api.Message;
+import org.opennms.core.ipc.sink.api.MessageConsumerManager;
+import org.opennms.core.ipc.sink.api.SinkModule;
+import org.opennms.core.ipc.sink.common.AbstractMessageDispatcherFactory;
+import org.opennms.core.logging.Logging;
+import org.opennms.core.logging.Logging.MDCCloseable;
+import org.osgi.service.cm.ConfigurationAdmin;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.amazon.sqs.javamessaging.SQSConnection;
+import com.codahale.metrics.JmxReporter;
+
+/**
+ * A factory for creating AwsRemoteMessageDispatcher objects.
+ * 
+ * @author <a href="mailto:agalue@opennms.org">Alejandro Galue</a>
+ */
+public class AwsRemoteMessageDispatcherFactory extends AbstractMessageDispatcherFactory<String> {
+
+    /** The Constant LOG. */
+    private static final Logger LOG = LoggerFactory.getLogger(AwsRemoteMessageDispatcherFactory.class);
+
+    /** The AWS configuration. */
+    private final Properties awsConfig = new Properties();
+
+    /** The configuration administration object. */
+    private ConfigurationAdmin configAdmin;
+
+    /** The reporter. */
+    private JmxReporter reporter;
+
+    /** The SQS session. */
+    private Session session;
+
+    /** The SQS Connection. */
+    private SQSConnection connection;
+
+    /** The producers. */
+    private Map<String,MessageProducer> producers = new HashMap<>();
+
+    /* (non-Javadoc)
+     * @see org.opennms.core.ipc.sink.common.AbstractMessageDispatcherFactory#getModuleMetadata(org.opennms.core.ipc.sink.api.SinkModule)
+     */
+    @Override
+    public <S extends Message, T extends Message> String getModuleMetadata(final SinkModule<S, T> module) {
+        return AwsUtils.getQueueName(module);
+    }
+
+    /* (non-Javadoc)
+     * @see org.opennms.core.ipc.sink.common.AbstractMessageDispatcherFactory#dispatch(org.opennms.core.ipc.sink.api.SinkModule, java.lang.Object, org.opennms.core.ipc.sink.api.Message)
+     */
+    @Override
+    public <S extends Message, T extends Message> void dispatch(SinkModule<S, T> module, String topic, T message) {
+        try (MDCCloseable mdc = Logging.withPrefixCloseable(MessageConsumerManager.LOG_PREFIX)) {
+            LOG.trace("dispatch({}): sending message {}", topic, message);
+
+            try {
+                TextMessage msg = session.createTextMessage(module.marshal((T)message));
```

Comment:
This method should be thread safe. I don't think that either the `session` or the `producer` can be safely used among threads.

Reply:
You're right, the Session is not thread safe. I'll rework that part.
I'm wondering, where is the documentation that says that everything used inside the `dispatch` has to be thread-safe ? Are there other methods with these kind of restrictions in this or another required class ?